### PR TITLE
Add new gosh shell and tests

### PR DESCRIPTION
**Pull request by MDr164**
This new shell aims to replace and deprecate the rush shell, ash and pkg/complete and pogosh. It uses mvdan/sh as an interpreter that understands mksh, bash and posix and uses c-bata/go-prompt for tab completion. Size wise it is roughly as small as the old rush shell was and still slightly smaller than the old in-tree elvish fork that existed. It's not as fully featured as the current elvish shell we use but it's much smaller and ideal for size constrained environments.

Signed-off-by: Marvin Drees <marvin.drees@9elements.com>

### CODE REVIEW

**Review by rminnich — DISMISSED** [review]

**Comment by rminnich** [comment]
can you, also, approve my PR to remove ash and pkg/complete?

**Comment by MDr164** [comment]
> can you, also, approve my PR to remove ash and pkg/complete?

Done and rebased this PR on the main again

File: go.mod [hunk 1]
```diff
@@ -33,28 +33,38 @@ require (
...
 	golang.org/x/text v0.3.3
 	golang.org/x/tools v0.1.1
 	gopkg.in/yaml.v2 v2.2.8
 	pack.ag/tftp v1.0.1-0.20181129014014-07909dfbde3c
 	src.elv.sh v0.16.3
 )
 
+require (
+	github.com/c-bata/go-prompt v0.2.6
+	mvdan.cc/sh/v3 v3.4.1
```
Comment (hugelgupf) [inline go.mod:47]: can you merge all 3 require blocks? i'm not even sure why they're separate

File: cmds/exp/gosh/gosh.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+	prompt "github.com/c-bata/go-prompt"
+	"github.com/c-bata/go-prompt/completer"
+	"golang.org/x/term"
+
+	"mvdan.cc/sh/v3/interp"
+	"mvdan.cc/sh/v3/syntax"
+)
+
+var tabCompletion = flag.Bool("tabcomp", false, "Enable tab completion.")
```
Comment (hugelgupf) [inline cmds/exp/gosh/gosh.go:25]: false by default?

File: cmds/exp/gosh/gosh_test.go [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+				t.Errorf("Unexpected error: %v", err)
+			} else if tt.wantErr != "" && fmt.Sprint(err) != tt.wantErr {
+				t.Errorf("Want error %q, got: %v", tt.wantErr, err)
+			}
+		})
+	}
+}
+
+func TestRunInteractiveTabCompletion(t *testing.T) {
+	// The function which should be tested here hardcodes
```
Comment (hugelgupf) [inline cmds/exp/gosh/gosh_test.go:298]: Could you test it by passing a test-only completerFunc?

File: go.mod (hunk 1 above)
Comment (MDr164) [reply go.mod:47]: go mod tidy made them seperate, I also wondered why but as I wasn't creating them manually I just left them as is. I'll manually change it back then.

File: cmds/exp/gosh/gosh.go (hunk 2 above)
Comment (MDr164) [reply cmds/exp/gosh/gosh.go:25]: Ah that was a leftover from testing, I think we can use tab completion by default so I'll change up the logic here.

File: cmds/exp/gosh/gosh_test.go (hunk 3 above)
Comment (MDr164) [reply cmds/exp/gosh/gosh_test.go:298]: I think it wasn't the completerFunc but rather the prompt.Input function that hardcodes device nodes internally. So I would have to create a test-only Input function in this case. That would require quite some refactoring but I could do so.

**Comment by rminnich** [comment]
oh darn, there is a comflict.

**Review by walterchris — APPROVED** [review]

**Review by rminnich — APPROVED** [review]

File: go.mod (hunk 1 above)
Comment (rminnich) [reply go.mod]: As with gofmt, I'm more inclined to let go mod tidy guide formatting?

**Pull request merged**